hs_core/xmlparser.py

```diff
@@ -0,0 +1,61 @@
+import base64
+import os.path
+from StringIO import StringIO
+from xml.sax import xmlreader, saxutils
+from xml.sax.expatreader import ExpatParser
+
+from django.conf import settings
+
+
+XML_CACHE = os.path.join(settings.PROJECT_ROOT, '_xmlcache')
+
+
+class CatalogedExpatParser(ExpatParser):
+    def external_entity_ref(self, context, base, sysid, pubid):
+        if not self._external_ges:
+            return 1
+
+        source = self._ent_handler.resolveEntity(pubid, sysid)
+        source = saxutils.prepare_input_source(source,
+                                               self._source.getSystemId() or
+                                               "")
+
+        # If an entry does not exist in the xml cache, create it.
+        filepath = os.path.join(XML_CACHE, base64.urlsafe_b64encode(pubid))
+        if not os.path.isfile(filepath):
+            with open(filepath, 'w') as f:
+                contents = source.getByteStream().read()
+                source.setByteStream(StringIO(contents))
+                f.write(contents)
```

Comment:
@jbradberry I am not sure how this pubid is being passed in to point to the cached DTD file (i.e., ```LS8vRFVCTElOIENPUkUvL0RDTUVTIERURCAyMDAyLzA3LzMxLy9FTg==```). I guess the intent is to only create the cached file when it is not there just to be safe, and we want to refer to the cached DTD file which is included as part of HydroShare codebase so that there is no need to contact external server for the DTD file. Is my understanding correct? This is the only thing I am not clear how it is done, otherwise, it looks good to me.